Added Temperature Sensor Support

hpmax opened the pull request:
Let's try this again from a fresh starting point.

This pull request adds support for the DS18B20 temperature sensor, which was already supported in hardware, but not software.

It adds two parameters in the sensord.conf:
output_POV_T which turns on the sentences (just like the other output parameters)
temp_config <9|10|11|12> This selects how many bits the user wants to use.  9 bits gives .5 degree  C resolution, 12 bits gives .0625 degree C resolution.  Conversion time is inversely proportional to resolution.  TANSTAAFL.

The code assumes that only a single DS18B20 is attached to the one-wire interface.  If more (or different devices) are plugged in, bad behavior may result.  Code is included, but is unused and untested to support multiple devices.

hpmax committed:
Added Temperature Sensor Support

linuxianer99 commented on main.c:
```diff
@@ -245,9 +247,41 @@ int NMEA_message_handler(int sock)
+
+	       // Send NMEA string via socket to XCSoar
+	       // send complete sentence including terminating '\0'
+	       if ((sock_err = send(sock, s, strlen(s)+1, 0)) < 0) fprintf(stderr, "send failed\n");
```
Think the NMEA string should not terminate with '/0'.
@kedder think this will revert your last PR, right ??

kedder replied:
Yeah, this is not right, but it is fixed in the next commit (that would be nice to squash into this one)

linuxianer99 commented on main.c:
```diff
@@ -245,9 +247,41 @@ int NMEA_message_handler(int sock)
+	int result;
+	int sock_err = 0;
+	
+	if (temp_sensor.active) {
```
Think it makes no sense to measure temp every sensor sample cycle. Think this break timing, isn't ?
You cna have a look in branch ds18b20. There wa already a sample implemantation for temp sensor.

hpmax replied:
Timo,

I think you are looking at the old version of the code.  But either way, the code works similar to the pressure measurement code.  I have an internal counter.  Every 12.5ms, it calls the temperature_measurement_handler which increments the counter.  On count 0 it instructs a conversion.  After some number of counts (which depends on the resolution) it reads out the temperature and then rolls over the counter.

So its not attempting to read out the temperature every cycle.  There is a capability to check when the conversion is done, which I'm not doing, but am considering adding.

I'll look at the ds18b20 branch.  I didn't even know it existed when I did this.

linuxianer99 commented on main.c:
```diff
@@ -245,9 +247,41 @@ int NMEA_message_handler(int sock)
+	     temp_counter=0;
+	     if ((temp_sensor.valid==1) && (config.output_POV_T == 1 )) {
+	       // Compose POV slow NMEA sentences
+	       result = Compose_Temperature_POV(&s[0], temp_sensor.temperature);
```
Think the NMEA sentence should be composed in NMEA_message_handler() but not in pressure_measurement_handler() as all other NMEA sentences

hpmax replied:
I looked over the ds2482 branch and while there are things I like about it more than my version (its definitely a little cleaner), there are also things I don't like about it -- in fact, based on my understanding of how the DS18B20 works, I'm surprised it works at all.

First, there is limited if any error checking.  My version could be improved, and I'd like to do that, but the ds2482 branch is worse.

Second, from what I can tell, it executes the pressure_measurement_handler function every 12.5ms.  On every 4th time through (50ms) it launches a temperature conversion.  It reads out the temperature measurement the next time through (12.5ms after the conversion request).  The fastest conversion time is 93.75ms, so it's always going to be initiating a new conversion before the previous conversion completes, and is always reading it out before valid data is available.

Mine on the other hand allows you to configure how much resolution you want (a feature the ds2482 branch seems to lack) and then waits the appropriate amount of time before reading out the measurement and launching a new conversion.  The reason I'm composing and writing out the NMEA sentence in temperature_measurement_handler is because I am composing the sentence AFTER valid data is available.  

Also, in my new version, I only do anything with the ds2482 if the config_POV_T switch is set.  I'm not sure why we would ever want to collect information from the sensors but not output it.

kedder commented on nmea.c:
```diff
@@ -183,6 +183,52 @@ int Compose_Voltage_POV(char *sentence, float voltage)
+  length = sprintf(sentence, "$POV,T,%+05.4f", temperature); 
+
+  // Calculate NMEA checksum and add to string
+  sprintf(sentence + length, "*%02X\n", NMEA_checksum(sentence));
```
NMEA sentences should end with `\r\n`

hpmax replied:
Okay, will fix that.

kedder commented on main.c:
```diff
@@ -548,6 +582,17 @@ int main (int argc, char **argv) {
+
+		if (ds2482_open(&temp_sensor,0x18) != 0)
+		{
+		  fprintf(stderr,"DS2482 sensor failed !!\n");
```
Would it be reasonable to ask to keep indentation style consistent? It seems like this patch has has it all: tabs, 8 spaces, 4 spaces, 2 spaces. It is very annoying to reconfigure your editor for each and every file you open when working on such an inconsistent codebase.

Right now the code in master is pretty consistent WRT to indentation - only tabs, no spaces. It would be nice to keep it that way.

hpmax replied:
Let me try to do that.

kedder commented on nmea.c:
```diff
@@ -183,6 +183,52 @@ int Compose_Voltage_POV(char *sentence, float voltage)
+
+  // check voltage input value for validity
+  if ((temperature < -55) || (temperature > 125)) {
+    temperature = 23;
```
Hm, why `23`?

hpmax replied:
I was just looking for a "sane" value.  I return a value of 0, and it should never get to this step if the temperature is invalid to begin.  I can double check that... But I don't think this is an issue.

hpmax replied:
The function was literally copied from the Compose Voltage POV function.  To be consistent, temp should probably be something out of range ( -100? ) and success should be 10 (I assumed that was a typ-o on my part, but looking back, it looks like it actually said 10.  Maybe this is a bug in the original code?

hpmax replied:
On second thought... if it doesn't get output, it doesn't matter... if it does get output... There may be good reason not to set it to something like -100.  The voltage reading of 0 is just for display, and is clearly bogus.  A temperature reading may be used in various atmospheric calculations and putting in bogus information may be a lot worse than putting in sane values like 23.  Thoughts?

kedder replied:
I agree, I think we should just skip it and not output anything if it is bogus. It should be better than just putting an outright lie.

kedder commented on nmea.c:
```diff
@@ -183,6 +183,52 @@ int Compose_Voltage_POV(char *sentence, float voltage)
+  length = sprintf(sentence, "$POV,T,%+05.4f", temperature); 
+
+  // Calculate NMEA checksum and add to string
+  sprintf(sentence + length, "*%02X\n", NMEA_checksum(sentence));
```
Should we actually output the sentence if `success` is 0?

hpmax replied:
See comment above.... This is modeled after the code that was already there.

kedder commented on main.c:
```diff
@@ -245,9 +247,41 @@ int NMEA_message_handler(int sock)
+	      OWWriteByte(&temp_sensor,0xCC); // Skip ROM
+	      OWWriteByte(&temp_sensor,0x44); // Conversion
+	      temp_counter=1;
+	   } else 
```
If I understand correctly, a function named "pressure measurement handler" is dealing with temperature sensor. Do we really need to put this code here? This function was already pretty large. If we can separate temperature and pressure measurement, it would be better to put them to separate functions.

hpmax replied:
Okay, will do that...

linuxianer99 replied:
Hmmm ... 
@kedder:  You are right with the naming, but the pressure measurement handler does all the timing of the measurement ... 
So we can create a own temperature handler, but should check the timing of the kalman filter etc... as there is more code in the loop now ...
Or do you think this timing is no problem at all ??

hpmax replied:
While I can't say that the temperature sensor didn't just make things worse, the current setup is a bit questionable if you are relying on accurate timing.  In fact, if you google usleep(), you'll see warnings galore not to rely on it for accurate timing and to use nanosleep() instead..  The current setup essentially drives a state machine whose state is updated 12.5ms after resolving the last state.  The problem, is that not all states take the same amount of time, so the delay between states is not constant.  Ideally, we'd probably want something like this to be interrupt driven, or at least something along those lines, and then do the pressure sensor stuff first.  If something like this is truly important for the Kalman filter then we probably want to get a better solution regardless of whether the temp sensor is included.

hpmax committed:
General cleanup

hpmax commented:
I'm still struggling to figure out how to fix the indent problem -- also, this has enough changes in it that I would consider it untested... Please do NOT approve this yet.  I will test this weekend and confirm everything is working, but I believe I have addressed all the other issues.  This version incorporates more and better error checking, return values are more consistent (fail is 0, or in the case of OWRead/OWrite is -1).

It now only operates the temperature sensor at all if you turn on the POV_T configuration, and does not report any value if a fault is encountered.  The POV string was also changed to have \r\n instead of \n.

kedder commented:
Ok, I think I can fix the indent problem in couple of minutes, so if you feel that's a lot of work for you, don't bother.  I'll submit a PR with reformat next.

hpmax commented:
Thanks for the offer.  In the short term, I'd appreciate it.  In the long term, I don't want to be dependent on you to fix my mess.  I'll install Atom and try to figure out how to deal with this.  As it stands, in emacs, I'm not even seeing the problem you're describing (although I'm sure its there).

However, please wait before making the changes.  I'd like to try adding one additional feature (to check the status of the conversion) and also test it since it's now been significantly re-written since what I originally tested.  I'll try to have everything done by the end of the weekend.

kedder commented:
Sure, there's no hurry, take your time.

hpmax committed:
Minor bug fixes

hpmax committed:
Minor bug fixes

hpmax commented:
The latest commit fixes all the bugs I'm aware of and it now works as designed.

However...

1) I do not think the timing is reliable... and it is equally as unreliable in the previous version.  If this is really timing sensitive... It opens the question up whether its actually important that the samples are collected at regular intervals or if its adequate for it to be at close to regular intervals and the exact time between samples is known, since this is MUCH easier to guarantee.

Looking at the Kalman code, it appears that there is a delta time input on the update function.  We could measure actual delta time and pass it in.

Immediately prior to executing the conversion for tep_sensor.p, we could execute the following:

clock_gettime(CLOCK_REALTIME,&timestruc);

Immediately after, you would execute:

deltatime=(timestruc.tv_nsec-oldtime)*1e-9;
oldtime=timestruc.tv_nsec;

This should be about as accurate as we can get.  That said, is there a reason why the Kalman filter only appears to be using ONE sensor, and not all of the available sensors?

2) The DS18B20 is acting weird.  The documentation indicates a 12-bit conversion should take a max of 750 ms, but my conversions seem to take less than 10ms.  

I'm going to try to contact Maxim and figure this out.
